ivy/functional/frontends/paddle/nn/functional/vision.py

```diff
@@ -135,3 +136,169 @@ def affine_grid(theta, out_shape, align_corners=True):
+        )  # size 2 representing (x,y) coordinates.
+
+    if mode == "nearest":
+        return grid_sample_nearest(input, grid, padding_mode)
+    elif mode == "bilinear":
+        return grid_sample_bilinear(input, grid, padding_mode)
+
+
+def grid_sample_nearest(input, grid, padding_mode):
+    """
+    Assigns the pixel value at non-integer grid coordinates to be the value of the
+    nearest pixel of the input data. The transformation process involves normalising the
+    grid & rounding them to the nearest pixel value.
+
+    parameters:
+    - input: image that needs to be transformed.
+    - grid: the transformation grid.
+    - padding_mode: controls how to handle out-of-bound values.
+
+    Returns:
+    transformed values for each point in the output grid.
+    These transformed values represent the result of applying spatial transformation.
+    """
+
+    # get the spatial dims of the input & grid
+    input_shape = ivy.shape(input)
+    ivy.shape(grid)
+    batch, channels, in_height, in_width = input_shape
+    # out_height, out_width = grid_shape[1], grid_shape[2]
+
+    # normalise grid coordinates to the range [-1, 1]
+    normalised_grid = 2.0 * grid / ivy.array([in_width - 1, in_height - 1])
+    normalised_grid = normalised_grid - ivy.array([1.0, 1.0])
+
+    # map the normalised grid coordinates to the output tensor using round operation
+    if padding_mode == "zeros":
+        grid_x = ivy.clip(math.round(normalised_grid[:, :, :, 0]), -1.0, 1.0)
+        grid_y = ivy.clip(math.round(normalised_grid[:, :, :, 1]), -1.0, 1.0)
+    else:
+        grid_x = ivy.clip(math.round(normalised_grid[:, :, :, 0]), 0.0, 1.0)
+        grid_y = ivy.clip(math.round(normalised_grid[:, :, :, 0]), 0.0, 1.0)
+
+    # convert the grid coordinaates to indices
+    x_indices = ((grid_x + 1.0) * 0.5 * (in_width - 1)).astype("int32")
+    y_indices = ((grid_y + 1.0) * 0.5 * (in_height - 1)).astype("int32")
+    # Gather values from input using indices
+    gathered = input[:, :, y_indices, x_indices]
+
+    return gathered
+    # the shape of the gathered tensor is (batch, channels, out_height, out_width)
+
+
+def grid_sample_bilinear(input, grid, padding_mode):
+    """
+    The bilinear interpolation calculates the pixel value at a non-integer grid
+    coordinate as a weighted average of the pixel values of the four nearest
+    neighbouring pixels. The interpolation is performed based on the relative distances
+    of the grid coordinates to these neighbouring pixels. Pixels closer to the grid
+    coordinate have higher weights, while pixels farther away have lower weights.
+
+    Parameters:
+    - input: data that needs to be transformed.
+    - grid: defines the transformation to be applied to the input tensor.
+    - padding_mode: controls how to handle out-of-bounds values.
+
+    Returns:
+    The transformed values for each point in the output grid.
+    These values represent the result of applying spatial transformation.
+    """
```

Comment:
please remove the docstring and rename the function to reflect it's private nature